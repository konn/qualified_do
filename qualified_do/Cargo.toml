[package]
name = "qualified_do"
version = "0.0.0"
license = "MIT"
edition = "2021"

[dependencies]
qualified_do_macro.workspace = true
functo_rs.workspace = true
itertools.workspace = true

[dev-dependencies]
either.workspace = true
