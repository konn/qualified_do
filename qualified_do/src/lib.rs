pub use qualified_do_macro::qdo;

pub mod iter;
pub use iter::Iter;
pub use iter::ZipIter;

pub use functo_rs::control::AsControl;
pub use functo_rs::data::AsData;
pub use functo_rs::impls::*;
pub use functo_rs::nonlinear::AsNonlinear;

pub type Optioned = AsControl<OptionFunctor>;
pub type Resulted<E> = AsControl<ResultFunctor<E>>;

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_optioned_applicative() {
        let answer = qdo! {Optioned {
            x <- Some(1);
            y <- Some(2);
            return x + y + 100
        }};
        assert_eq!(answer, Some(103));
    }

    #[test]
    fn text_optioned_resulted_nested() {
        #[derive(Debug, Copy, Clone)]
        enum Go {
            Go,
            NoGo,
        }
        let ans: fn(Go) -> Result<i64, String> = |go: Go| {
            qdo! { Resulted {
                x <- qdo!{ Optioned {
                    x <- Some(1);
                    y <- Some(2);
                    Go::Go <- Some(go);
                    guard x + y % 2 == 1;
                    return x + y + 100
                }}.ok_or("Failed".to_string());
                y <- Ok(3);
                return x + y + 1000
            }}
        };
        assert_eq!(ans(Go::Go), Ok(1106));
        assert_eq!(ans(Go::NoGo), Err("Failed".to_string()));
    }
}
