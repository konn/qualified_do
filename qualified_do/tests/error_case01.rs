fn main() {
    use qualified_do::qdo;

    qdo! {hoge::fuga {}}
}
