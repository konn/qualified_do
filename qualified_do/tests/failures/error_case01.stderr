error: expected at least one statement
 --> tests/error_case01.rs:4:23
  |
4 |     qdo! {hoge::fuga {}}
  |                       ^
